src/app.cpp

```diff
@@ -94,14 +117,28 @@ void transform(
 		if (header.data_type == TECMP_DATA_CAN || header.data_type == TECMP_DATA_CANFD) {
 			struct canfd_frame can = { 0 };
 			can.can_id = ntoh32(*((uint32_t*)data));
-			can.len = data[4];
 			if (header.data_type == TECMP_DATA_CANFD)
 			{
 				can.flags |= CANFD_FDF;
 				can.flags |= header.data_flags & TMP_BITRATE_SWITCH ? CANFD_BRS : 0;
 				can.flags |= header.data_flags & TMP_ERROR_NODE_ACTIVE ? CANFD_ESI : 0;
 			}
-			memcpy(can.data, data + 5, can.len);
+			if ((header.data_flags & (TMP_ERROR_MESSAGE | TMP_BITSTUFF_ERROR |
+				TMP_ACK_DEL_ERROR | TMP_CRC_DEL_ERROR | TMP_CRC_ERROR | TMP_EOF_ERROR)) != 0) {
+				can.len = 8;
+				// If the data type is CAN-FD, force the frame to be a Classical CAN error frame
+				bool is_canfd = header.data_type == TECMP_DATA_CANFD;
+
+				if (is_canfd) {
+					const uint8_t CANFD_TO_CAN_ERROR_MASK = 0x02;
+					can.flags &= CANFD_TO_CAN_ERROR_MASK;
```

Comment:
What are you doing here?

Reply:
I used this mask to set the bit related to the **"Error State Indicator" (ESI)** to `1` but when this bit is set to `1`, the frame is interpreted as a **normal frame**, not an **error frame**. To address this, I set the **"FD Frame"** flag to `false`. From what I understand, a **CAN error frame** should always be treated as a standard CAN frame, even if the original frame over the TECMP is CAN FD.
![image](https://github.com/user-attachments/assets/b3ed1dfd-8b1e-43fc-bb5c-24a5bb76f945)

Reply:
`Error State Indicator` in an actual bit in the CAN FD frame, don't mix it up with Error Frames, they are not related.

Reply:
@kayoub5  Without setting the CAN-FD frame flag to false, the packet is treated as a CAN-FD frame without error detection!
![image](https://github.com/user-attachments/assets/1c1c7f94-4627-47bc-bfba-f581d10a1b8c)